kostal/integrationpluginkostal.cpp

```diff
@@ -0,0 +1,516 @@
+            kostalConnection->connectDevice();
+        } else if (!reachable) {
+            // Note: We disable autoreconnect explicitly and we will
+            // connect the device once the monitor says it is reachable again
+            kostalConnection->disconnectDevice();
+        }
+    });
+
+    connect(kostalConnection, &KostalModbusTcpConnection::reachableChanged, thing, [this, thing, kostalConnection](bool reachable){
+        qCDebug(dcKostal()) << "Reachable changed to" << reachable << "for" << thing;
+        if (reachable) {
+            // Connected true will be set after successfull init
+            kostalConnection->initialize();
+        } else {
+            thing->setStateValue("connected", false);
+            foreach (Thing *childThing, myThings().filterByParentId(thing->id())) {
+                childThing->setStateValue("connected", false);
+            }
+        }
+    });
+
+    connect(kostalConnection, &KostalModbusTcpConnection::initializationFinished, thing, [=](bool success){
+        if (!thing->setupComplete())
+            return;
+
+        if (success) {
+            thing->setStateValue("connected", true);
+        } else {
+            thing->setStateValue("connected", false);
+            // Try once to reconnect the device
+            kostalConnection->reconnectDevice();
+        }
+    });
+
+    connect(kostalConnection, &KostalModbusTcpConnection::initializationFinished, info, [=](bool success){
+        if (!success) {
+            qCWarning(dcKostal()) << "Connection init finished with errors" << thing->name() << kostalConnection->hostAddress().toString();
+            hardwareManager()->networkDeviceDiscovery()->unregisterMonitor(monitor);
+            kostalConnection->deleteLater();
+            info->finish(Thing::ThingErrorHardwareFailure, QT_TR_NOOP("Could not initialize the communication with the inverter."));
+            return;
+        }
+
+        qCDebug(dcKostal()) << "Connection init finished successfully" << kostalConnection;
+        m_kostalConnections.insert(thing, kostalConnection);
+        info->finish(Thing::ThingErrorNoError);
+
+        // Set connected true
+        thing->setStateValue("connected", true);
+        foreach (Thing *childThing, myThings().filterByParentId(thing->id())) {
+            childThing->setStateValue("connected", true);
+        }
+
+        // Make sure we use the same endianness as the inverter (register size 1, so independent from endianess)
+        switch (kostalConnection->modbusByteOrder()) {
+        case KostalModbusTcpConnection::ByteOrderBigEndian:
+            kostalConnection->setEndianness(ModbusDataUtils::ByteOrderBigEndian);
+            break;
+        case KostalModbusTcpConnection::ByteOrderLittleEndian:
+            kostalConnection->setEndianness(ModbusDataUtils::ByteOrderLittleEndian);
+            break;
+        }
+
+        connect(kostalConnection, &KostalModbusTcpConnection::updateFinished, thing, [=](){
```

Comment:
If I'm reading it correctly, this isn't good because
* the KostalConnection is created
* onReachableChanged has a slot connected that calls initialize()
* initialize() will trigger onInitialized
* onInitialized connects slots to the connection

=> Every time the wallbox becomes reachable false -> true, the slots will be connected once more.

Reply:
On reachable changed will trigger an initialization. If the initialization succeeded, the setup is finished and/of the thing is connected.

There are to slots connected to initialization finished. The first one is for the lifetime of the thing handling the connected state. If init failed, reconnect once.

    connect(kostalConnection, &KostalModbusTcpConnection::initializationFinished, thing, [=](bool success){
        if (!thing->setupComplete())
            return;
        ....

The second is only for the setup. This connection works only as long as the `info` exists. On finished, all connections will be done once and we are done with the setup. 

    connect(kostalConnection, &KostalModbusTcpConnection::initializationFinished, info, [=](bool success){

I see no issue here. Can you verify after the explanation?

Reply:
right, my bad. it's ok...

Wouldn't the initializationFinished connection always reconnect until it eventually succeeds (Even through the comment says "once")?

Reply:
if it does not succeed, the connection gets deleted, finishes the setup with an error and clean up. On the next setup, once it succeeds, it will create all connections once. From the moment on where the setup has completed, the upper connection takes care about the init handling.